sympy/simplify/cse_diff.py

```diff
@@ -0,0 +1,188 @@
+            The vector to be differentiated.
+
+        wrt : Matrix, list, or tuple
+            The vector with respect to which to do the differentiation. Can be a matrix or an iterable of variables.
+
+        See Also
+        ========
+
+        Direct Acyclic Graph : https://en.wikipedia.org/wiki/Directed_acyclic_graph
+
+    """
+
+    # Check Inputs
+    if not isinstance(wrt, (MatrixBase, list, tuple)):
+        raise TypeError("``wrt`` must be an iterable of variables")
+
+    elif isinstance(wrt, (list, tuple)):
+        wrt = Matrix(wrt)
+
+    if not isinstance(expr, MatrixBase):
+        raise TypeError("``expr`` must be of matrix type")
+
+    # Both ``wrt`` and ``expr`` can be a row or a column matrix, so we need to make
+    # sure all valid combinations work, but everything else fails:
+    if not (expr.shape[0] == 1 or expr.shape[1] == 1):
+        raise TypeError("``expr`` must be a row or a column matrix")
+
+    if not (wrt.shape[0] == 1 or wrt.shape[1] == 1):
+        raise TypeError("``wrt`` must be a row or a column matrix")
+
+    replacements, reduced_expr = cse(expr)
+    replacements, reduced_expr = postprocess(replacements, reduced_expr)
+
+    if replacements:
+        rep_sym, sub_expr = map(Matrix, zip(*replacements))
+    else:
+        rep_sym, sub_expr, reduced_expr = Matrix([]), Matrix([]), [expr]
+
+    l_sub, l_wrt, l_red = len(sub_expr), len(wrt), len(reduced_expr[0])
+
+    f1 = {
+        (i, j): diff_value
+        for i, r in enumerate(reduced_expr[0])
+        for j, w in enumerate(wrt)
+        if (diff_value := r.diff(w)) != 0
+    }
+
+    if not replacements:
+        return SparseMatrix(l_red, l_wrt, f1)
+
+    f2 = {
+        (i, j): diff_value
+        for i, (r, fs) in enumerate([(r, r.free_symbols) for r in reduced_expr[0]])
+        for j, s in enumerate(rep_sym)
+        if s in fs and (diff_value := r.diff(s)) != 0
+    }
+
+    symbols = expr.free_symbols
+    precomputed_fs = [s.free_symbols - symbols for s in sub_expr]
+
+    C = Counter({(0, j): diff_value for j, w in enumerate(wrt) if (diff_value := sub_expr[0].diff(w)) != 0})
+
+    for i in range(1, l_sub):
+        Bi = {(i, j): diff_value for j in range(i + 1)
+              if rep_sym[j] in precomputed_fs[i] and (diff_value := sub_expr[i].diff(rep_sym[j])) != 0}
+
+        Ai = Counter({(i, j): diff_value for j, w in enumerate(wrt)
+                      if (diff_value := sub_expr[i].diff(w)) != 0})
+
+        if Bi:
+            Ci = dok_matrix_multiply(Bi, C)
+            Ci.update(Ai)  # Use Counter's update method to add Ai to Ci
+            C.update(Ci)  # Update C with the result
+        else:
+            C.update(Ai)
+
+    J = dok_matrix_multiply(f2, C)
+
+    for (i, j), value in f1.items():
+        J[(i, j)] += value
+
+    sub_rep = dict(replacements)
+    for i, ik in enumerate(precomputed_fs):
+        sub_dict = {j: sub_rep[j] for j in ik}
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
```

Comment:
It would be nice to have a function that returns the cse'd Jacobian and then a second function that is a light wrapper that subs back in the originals to give the full Jacobian.